src/benchmarks.cc

```diff
@@ -0,0 +1,289 @@
+#include <mpi.h>
+#include <Kokkos_Random.hpp>
+#include <Kokkos_Complex.hpp>
+
+#include "ops.h"
+#include "benchmarks.h"
+
+#include <iostream>
+
+namespace benchmarks {
+
+template <>
+std::string typeToString<double>() {
+    return "double ";
+}
+
+template <>
+std::string typeToString<Kokkos::complex<double>>() {
+    return "complex";
+}
+
+std::string benchmarkToString(const benchmark_types& b) {
+    switch (b) {
+        case level1: return "level1";
+        case level2: return "level2";
+        case level3: return "level3";
+        case dpotrf: return "dpotrf";
+        default: return "unknown";
+    }
+}
+
+template <typename T>
+benchmark_results_t runBenchmarkLevel1(int N, int iters) {
+    Kokkos::View<T*> x("x", N);
+    Kokkos::View<T*> y("y", N);
+
+    T* x_ptr = x.data();
+    T* y_ptr = y.data();
+
+    Kokkos::Random_XorShift64_Pool pool(123);
+    Kokkos::fill_random(x, pool, 10.0);
+    Kokkos::fill_random(y, pool, 10.0);
+
+    std::vector<double> iter_timings;
+    double total_time = 0.0;
+
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    for (int i = 0; i < iters; i++) {
+        Kokkos::Timer timer;
+        ops::level1<T>(N, x_ptr, y_ptr);
+        Kokkos::fence();
+
+        // Skip the first iteration
+        if (i > 0) {
+            double time = timer.seconds();
+            total_time += time;
+            iter_timings.push_back(time);
+        }
+    }
+
+    int rank = -1;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    std::cout << "[level1 " << typeToString<T>() << "] (N=" << N << ") rank: " << rank << ", total_time=" << total_time << std::endl;
+
+    return std::make_tuple(iter_timings, total_time);
+}
+
+template <typename T>
+benchmark_results_t runBenchmarkLevel2(int M, int N, int iters) {
+    Kokkos::View<T*> x("x", M);
+    Kokkos::View<T*> y("y", N);
+    Kokkos::View<T**> A("A", M, N);
+
+    T* x_ptr = x.data();
+    T* y_ptr = y.data();
+    T* A_ptr = A.data();
+
+    Kokkos::Random_XorShift64_Pool pool(123);
+    Kokkos::fill_random(x, pool, 10.0);
+    Kokkos::fill_random(y, pool, 10.0);
+    // Kokkos::fill_random(A, pool, 10.0);
```

Comment:
Why is this commented out?

Reply:
Fixed this and unified all output buffers so they are touched *before* the kernel starts